File: src/routes/assessment-progress/assessment-progress.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  HttpCode,
  HttpStatus,
  UseGuards,
  ParseIntPipe,
  Request,
} from '@nestjs/common'
import { AssessmentProgressService } from './assessment-progress.service'
import {
  SaveAnswerProgressDTO,
  UpdateAnswerProgressDTO,
  QueryAssessmentProgressDTO,
  SubmitAssessmentDTO,
  StartAssessmentDTO,
  AutoSaveProgressDTO,
} from './assessment-progress.dto'
import { Auth } from 'src/shared/decorators/auth.decorator'
import { AuthType } from 'src/shared/constants/auth.constant'
import { Roles } from 'src/shared/decorators/roles.decorator'
import { RolesGuard } from 'src/shared/guards/roles.guard'
import { RoleName } from 'src/shared/constants/role.constant'
import { ActiveUser } from 'src/shared/decorators/active-user.decorator'

@Controller('assessment-progress')
@UseGuards(RolesGuard)
export class AssessmentProgressController {
  constructor(private readonly progressService: AssessmentProgressService) {}

  @Post('start')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.CREATED)
  async startAssessment(@Body() startDto: StartAssessmentDTO, @ActiveUser('userId') userId: number) {
    return this.progressService.startAssessment(startDto, userId)
  }

  @Post('submit')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async submitAssessment(@Body() submitDto: SubmitAssessmentDTO, @ActiveUser('userId') userId: number) {
    return this.progressService.submitAssessment(submitDto, userId)
  }

  @Post('answers')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.CREATED)
  async saveAnswer(@Body() saveDto: SaveAnswerProgressDTO, @ActiveUser('userId') userId: number) {
    return this.progressService.saveAnswer(saveDto, userId)
  }

  @Put('answers/:id')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async updateAnswer(
    @Param('id', ParseIntPipe) answerId: number,
    @Body() updateDto: UpdateAnswerProgressDTO,
    @ActiveUser('userId') userId: number,
  ) {
    return this.progressService.updateAnswer(answerId, updateDto, userId)
  }

  @Post('auto-save')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async autoSave(@Body() autoSaveDto: AutoSaveProgressDTO, @ActiveUser('userId') userId: number) {
    return this.progressService.autoSave(autoSaveDto, userId)
  }

  @Get('my')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getAllMyProgresses(@ActiveUser('userId') userId: number, @Query() queryDto: QueryAssessmentProgressDTO) {
    return this.progressService.getAllMyProgresses(userId, queryDto)
  }

  @Get('my-assignments')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getAllMyProgressesAssignment(
    @ActiveUser('userId') userId: number,
    @Query() queryDto: QueryAssessmentProgressDTO,
  ) {
    return this.progressService.getAllMyProgressesAssignment(userId, queryDto)
  }

  @Get('my/stats')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getMyStats(@ActiveUser('userId') userId: number) {
    return this.progressService.getUserStats(userId)
  }

  @Get('assessment/:assessmentId/my')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getMyProgress(@Param('assessmentId', ParseIntPipe) assessmentId: number, @ActiveUser('userId') userId: number) {
    return this.progressService.getMyProgress(assessmentId, userId)
  }

  @Get(':id')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getProgressById(@Param('id', ParseIntPipe) progressId: number, @ActiveUser('userId') userId: number) {
    return this.progressService.getProgressById(progressId, userId)
  }

  @Get(':id/answers')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async getMyAnswers(@Param('id', ParseIntPipe) progressId: number, @ActiveUser('userId') userId: number) {
    return this.progressService.getMyAnswers(progressId, userId)
  }

  @Delete(':id')
  @Auth([AuthType.Bearer])
  @HttpCode(HttpStatus.OK)
  async deleteProgress(@Param('id', ParseIntPipe) progressId: number, @ActiveUser('userId') userId: number) {
    return this.progressService.deleteProgress(progressId, userId)
  }

  // ============= ADMIN & STATS =============

  @Get()
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async getAllProgresses(@Query() queryDto: QueryAssessmentProgressDTO) {
    return this.progressService.getAllProgresses(queryDto)
  }

  @Get('assessment/:assessmentId/stats')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Admin, RoleName.Lecturer)
  @HttpCode(HttpStatus.OK)
  async getAssessmentStats(@Param('assessmentId', ParseIntPipe) assessmentId: number) {
    return this.progressService.getAssessmentStats(assessmentId)
  }
}

File: assessment-service/src/main/java/com/torii/assessment/controller/AssessmentProgressController.java
package com.torii.assessment.controller;

import com.torii.assessment.dto.progress.AssessmentAnswerProgressDTO;
import com.torii.assessment.dto.assessment.AssessmentDTO;
import com.torii.assessment.dto.progress.AssessmentProgressDTO;
import com.torii.assessment.dto.progress.AutoSaveProgressDTO;
import com.torii.assessment.dto.progress.SaveAnswerProgressDTO;
import com.torii.assessment.dto.progress.StartAssessmentProgressDTO;
import com.torii.assessment.dto.progress.SubmitAssessmentProgressDTO;
import com.torii.assessment.dto.progress.UpdateAnswerProgressDTO;
import com.torii.assessment.service.AssessmentProgressService;
import com.torii.assessment.util.RequestAuthUtil;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/assessment-progress")
@RequiredArgsConstructor
public class AssessmentProgressController {

    private final AssessmentProgressService assessmentProgressService;

    @PostMapping("/start")
    public ResponseEntity<AssessmentProgressDTO> startAssessment(
            @Valid @RequestBody StartAssessmentProgressDTO dto,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(assessmentProgressService.startAssessment(dto, authUser.userId()));
    }

    @PostMapping("/answers")
    public ResponseEntity<AssessmentAnswerProgressDTO> saveAnswer(
            @Valid @RequestBody SaveAnswerProgressDTO dto,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(assessmentProgressService.saveAnswerProgress(dto, authUser.userId(), authUser.role()));
    }

    @PutMapping("/answers/{id}")
    public ResponseEntity<AssessmentAnswerProgressDTO> updateAnswer(
        @PathVariable Long id,
        @RequestBody UpdateAnswerProgressDTO dto,
        HttpServletRequest request
    ) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.updateAnswerProgress(id, dto, authUser.userId(), authUser.role()));
    }

    @PostMapping("/auto-save")
    public ResponseEntity<AssessmentProgressDTO> autoSave(
            @Valid @RequestBody AutoSaveProgressDTO dto,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.autoSave(dto, authUser.userId(), authUser.role()));
    }

    @PostMapping("/submit")
    public ResponseEntity<AssessmentProgressDTO> submitAssessment(
            @Valid @RequestBody SubmitAssessmentProgressDTO dto,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.submitAssessment(dto, authUser.userId(), authUser.role()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<AssessmentDTO> getProgressById(
            @PathVariable Long id,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.getProgressDetailById(id, authUser.userId(), authUser.role()));
    }

    @GetMapping("/user")
    public ResponseEntity<List<AssessmentProgressDTO>> getUserProgresses(
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.getUserProgresses(authUser.userId()));
    }

    @GetMapping("/assessment/{assessmentId}/user/{userId}")
    public ResponseEntity<AssessmentProgressDTO> getUserProgressByAssessment(
        @PathVariable Long assessmentId,
        @PathVariable Integer userId,
        HttpServletRequest request
    ) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        RequestAuthUtil.ensureSelfOrPrivileged(authUser, userId);
        return ResponseEntity.ok(assessmentProgressService.getUserProgressByAssessment(assessmentId, userId));
    }

    @GetMapping("/{id}/answers")
    public ResponseEntity<List<AssessmentAnswerProgressDTO>> getAnswersByProgress(
            @PathVariable("id") Long progressId,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        return ResponseEntity.ok(assessmentProgressService.getAnswersByProgress(progressId, authUser.userId(), authUser.role()));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> deleteProgress(
            @PathVariable Long id,
            HttpServletRequest request) {
        RequestAuthUtil.AuthUser authUser = RequestAuthUtil.getAuthUser(request);
        assessmentProgressService.deleteProgress(id, authUser.userId(), authUser.role());
        return ResponseEntity.ok(Map.of("message", "Progress deleted successfully"));
    }
}
